pkg/kyverno/common/common.go

```diff
@@ -571,10 +572,11 @@ func ApplyPolicyOnResource(policy *v1.ClusterPolicy, resource *unstructured.Unst
 	}
 
 	mutateResponse := engine.Mutate(&engine.PolicyContext{Policy: *policy, NewResource: *resource, JSONContext: ctx, NamespaceLabels: namespaceLabels})
-	err = processMutateEngineResponse(policy, mutateResponse, resPath, rc, mutateLogPath, stdin, mutateLogPathIsDir, resource.GetName())
+	engineResponses = append(engineResponses, mutateResponse)
```

Comment:
`ApplyPolicyOnResource` would apply a policy on a resource. 
so when the policy gets mutated, it will generate a `mutateResponse`. `mutate` happens only once. 
Why are we retuning an array of `*response.EngineResponse` why not a single `*response.EngineResponse`

Reply:
Regarding `*response.EngineResponse` and  `[ ]*response.EngineResponse` what have I think is that   `ApplyPolicyOnResource` will return `[ ]*response.EngineResponse`, and it will contain responses of mutate, validate and later on generate. 

```
func ApplyPolicyOnResource () { 
    .
    .
    engineResponses := make([]*response.EngineResponse, 0)
    .
    .
    if mutateResponse != nil {
		    engineResponses = append(engineResponses, mutateResponse)
    }
    .
    .
    if validateResponse != nil {
		    engineResponses = append(engineResponses, validateResponse)
    }
    .
    .
    similarly we will add generateResponse too. 
    if generateResponse!= nil {
		    engineResponses = append(engineResponses, generateResponse)

}
```
I tried above way and it works..
What is your suggestion ??

Reply:
see [this](https://github.com/kyverno/kyverno/pull/2279#discussion_r714520471)